### ISSUE TITLE
[Windows] error in knitting process Rmd

### ISSUE THREAD

**Issue by RMHogervorst** [airbnb/knowledge-repo#109]
I modified step 5 
```
5. Add your post to the repo with path project/example

knowledge_repo --repo ./example_repo add example_post.ipynb -p project/example
```
To work with Rmd

`knowledge_repo --repo ./example_repo add example_post.Rmd -p project/example`

However I recieved the folowing error

```
Error: '\R' is an unrecognized escape in character string starting "'D:\R"
Execution halted
Traceback (most recent call last):
  File "C:\Users\roel\AppData\Local\Programs\Python\Python35-32\Scripts\knowledge_repo", line 248, in <module>
    kp = knowledge_repo.KnowledgePost.from_file(args.filename, src_paths=args.src)
  File "c:\users\roel\appdata\local\programs\python\python35-32\lib\site-packages\knowledge_repo\post.py", line 347, in from_file
    kp = KnowledgePostConverter.for_file(cls(), filename, format=format, postprocessors=postprocessors).from_file(filename, **opts)
  File "c:\users\roel\appdata\local\programs\python\python35-32\lib\site-packages\knowledge_repo\converter.py", line 43, in wrapped
    f(*args, **kwargs)
  File "c:\users\roel\appdata\local\programs\python\python35-32\lib\site-packages\knowledge_repo\converters\rmd.py", line 24, in from_file
    subprocess.check_output(runcmd, shell=True)
  File "c:\users\roel\appdata\local\programs\python\python35-32\lib\subprocess.py", line 626, in check_output
    **kwargs).stdout
  File "c:\users\roel\appdata\local\programs\python\python35-32\lib\subprocess.py", line 708, in run
    output=stdout, stderr=stderr)
subprocess.CalledProcessError: Command 'R --vanilla --slave -e "library(knitr); setwd('D:\RmhDocs\Documents\docs\actief\knowledge-repo');                         x = knit('D:\RmhDocs\Documents\docs\actief\knowledge-repo\example_post.Rmd', 'C:\Users\roel\AppData\Local\Temp\tmpwa3xe977', quiet=T)" ' returned non-zero exit status 1
````

I think somewhere in the script `knowledge_repo\converters\rmd.py`  the problem resides

**Comment by Jim89** [comment]
I think it's due to Windows and it's insistence on using '\' rather than '/' for file paths (e.g. see [this](http://stackoverflow.com/questions/8425409/file-path-issues-in-r-using-windows-hex-digits-in-character-string-error/27071930#27071930) SO post), coupled with `R`'s insistence on using `/`. Doesn't exactly make for a joyous situation.

Easiest solution is to replace all '\' with '\\' to escape them and make `R` happy.

**Comment by RMHogervorst** [comment]
There is actually a os independent solution with file.path() i believe

Op 4 nov. 2016 16:33 schreef "Jim Leach" <notifications@github.com>:

I think it's due to Windows and it's insistence on using '\' rather than
'/' for file paths (e.g. see this
<http://stackoverflow.com/questions/8425409/file-path-issues-in-r-using-windows-hex-digits-in-character-string-error/27071930#27071930>
SO post), coupled with R's insistence on using /. Doesn't exactly make for
a joyous situation.

Easiest solution is to replace all '\' with '\' to escape them and make R
happy.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub
<https://github.com/airbnb/knowledge-repo/issues/109#issuecomment-258464143>,
or mute the thread

**Comment by Jim89** [comment]
There is, but the way they call `R` and `knitr` to create the post from an Rmd is using the variables from the command line (see [here](https://github.com/airbnb/knowledge-repo/blob/master/knowledge_repo/converters/rmd.py)). Rather than embed a call to `file.path()` inside the in-line `R` command it's probably easier to do a find-and-replace on the variables first (perhaps only if the OS is detected to be Windows).

**Comment by RMHogervorst** [comment]
Yeah fair enough.

Op 4 nov. 2016 16:43 schreef "Jim Leach" <notifications@github.com>:

**Comment by Jim89** [comment]
Ok `os.path.abspath()` don't actually correctly escape `\` (it just looks that way when testing in my notebook - using `print()` it looks like it's producing a single `\`).

**Comment by RMHogervorst** [comment]
This is not yet solved, or does ` pip install knowledge-repo` not take the latest version from github?

Sorry, i see that it's using the version from november 3th [https://pypi.python.org/pypi/knowledge-repo/0.6.11]

**Comment by RMHogervorst** [comment]
I tried installing the latest version using github. But that didn't change teh problem.

**Comment by earthmancash** [comment]
@RMHogervorst Sorry that this fell through the cracks! Are you getting the same subprocess command from earlier in this issue still? If so, do you know what the `'R --vanilla --slave -e "library(knitr); setwd('D:\RmhDocs\Documents\docs\actief\knowledge-repo');                         x = knit('D:\RmhDocs\Documents\docs\actief\knowledge-repo\example_post.Rmd', 'C:\Users\roel\AppData\Local\Temp\tmpwa3xe977', quiet=T)" '` command should look like for a windows setup?

**Comment by tom-christie** [comment]
I'm running into this issue on MacOS 10.12.  I run into this error when I run the R command directly:

    Warning message:
    package ‘knitr’ was built under R version 3.2.5 
    Error in file(file, ifelse(append, "a", "w")) : 
      cannot open the connection
    Calls: knit ... parse_all -> .handleSimpleError -> h -> cat -> file
    In addition: Warning message:
    In file(file, ifelse(append, "a", "w")) :
      cannot open file '/tmp/tmpguhoqcyt': Permission denied
    Execution halted
    
On this OS the /tmp is actually a symlink to /private/tmp.  The /private/tmp directory has 777 permissions but /tmp doesn't.  I don't know whether this is related to the issue, but it seems like it might be.

**Comment by earthmancash** [comment]
@tom-christie can you tell me what exact R command you're using? Wondering what the path to your Rmd file looks like.

**Comment by tom-christie** [comment]
Sure thing - 

    >> R --vanilla --slave -e "library(knitr); setwd('/Users/USERNAME/git');  x = knit('/Users/USERNAME/git/example_post.Rmd', '/tmp/tmpguhoqcyt', quiet=T)"
    Warning message:
    package ‘knitr’ was built under R version 3.2.5 
    Error in file(file, ifelse(append, "a", "w")) : 
      cannot open the connection
    Calls: knit ... parse_all -> .handleSimpleError -> h -> cat -> file
    In addition: Warning message:
    In file(file, ifelse(append, "a", "w")) :
      cannot open file '/tmp/tmpguhoqcyt': Permission denied
    Execution halted

**Comment by matthewwardrop** [comment]
@RMHogervorst @Jim89 Is this still an issue? I plan to solve this within the next day or so :). It's about time!

@tom-christie It looks like your error is a different one. If you are still having trouble with it, can you open a new issue and I'll try to resolve it in the next few days!

**Comment by RMHogervorst** [comment]
checking today @matthewwardrop [2017-02-16]

**Comment by RMHogervorst** [comment]
I've updated knowledge repo with pip install  --upgrade knowledge-repo

But I get the following errors

(Ive replaced some paths)
```
PATHTOmydocuments \knowledge-repo>knowledge_repo --repo ./example_repo add example_post.Rmd -p project/example_rmd
INFO:knowledge_repo.repositories.gitrepository:Fetching updates to the knowledge repository...
Warning message:
package 'knitr' was built under R version 3.3.2
Traceback (most recent call last):
  File "C:\Users\roel\Miniconda3\Scripts\knowledge_repo", line 254, in <module>
    kp = knowledge_repo.KnowledgePost.from_file(args.filename, src_paths=args.src)
  File "c:\users\roel\miniconda3\lib\site-packages\knowledge_repo\post.py", line 350, in from_file
    kp = KnowledgePostConverter.for_file(cls(), filename, format=format, postprocessors=postprocessors).from_file(filename, **opts)
  File "c:\users\roel\miniconda3\lib\site-packages\knowledge_repo\converter.py", line 43, in wrapped
    f(*args, **kwargs)
  File "c:\users\roel\miniconda3\lib\site-packages\knowledge_repo\converters\rmd.py", line 38, in from_file
    os.remove(tmp_path)
PermissionError: [WinError 32] The process cannot access the file because it is being used by another process: 'C:\\Users\\roel\\AppData\\Local\\Temp\\tmpnpqwqbi9'
```


I might try a new installation of python or perhaps some virtual environment because it keeps thworing errors in my windows installation.

**Comment by matthewwardrop** [comment]
@RMHogervorst @Jim89 This should now be fixed in #228. Will close this issue once it is merged into master.
